src/Compilers/Core/Portable/Compilation/SymbolInfo.cs

```diff
@@ -91,35 +72,20 @@ internal SymbolInfo(ISymbol? symbol, ImmutableArray<ISymbol> candidateSymbols, C
             this.CandidateReason = candidateReason;
         }
 
+        internal ImmutableArray<ISymbol> GetAllSymbols()
+            => this.Symbol == null ? CandidateSymbols : ImmutableArray.Create(this.Symbol);
+
         public override bool Equals(object? obj)
-        {
-            return obj is SymbolInfo && Equals((SymbolInfo)obj);
-        }
+            => obj is SymbolInfo info && Equals(info);
 
         public bool Equals(SymbolInfo other)
-        {
-            if (!object.Equals(this.Symbol, other.Symbol) ||
-                _candidateSymbols.IsDefault != other._candidateSymbols.IsDefault ||
-                this.CandidateReason != other.CandidateReason)
-            {
-                return false;
-            }
-
-            return _candidateSymbols.IsDefault || _candidateSymbols.SequenceEqual(other._candidateSymbols);
-        }
+            => this.CandidateReason == other.CandidateReason &&
+               Equals(this.Symbol, other.Symbol) &&
+               CandidateSymbols.SequenceEqual(other.CandidateSymbols);
 
         public override int GetHashCode()
-        {
-            return Hash.Combine(this.Symbol, Hash.Combine(Hash.CombineValues(_candidateSymbols, 4), (int)this.CandidateReason));
-        }
+            => Hash.Combine(this.Symbol, Hash.Combine(Hash.CombineValues(_candidateSymbols, 4), (int)this.CandidateReason));
 
-        internal bool IsEmpty
-        {
-            get
-            {
-                return this.Symbol == null
-                    && this.CandidateSymbols.Length == 0;
-            }
-        }
+        internal bool IsEmpty => this.Symbol == null && this.CandidateSymbols.Length == 0;
```

Comment:
open question if this is correct.  For example, you could have a non-0 .CandidateReason.  That said, perhaps the constructor should throw if CandidateReason is non-0 and tehre are no CandidateSymbols.  We may also want to beef up the documentation here to make it clear about the relation between CandidateSymbols and CandidateReason.  We do say there's a CandidateReason if CandidateSymbols is non-empty.  But we dont' say that if there's a CandidateReason that CandidateSymbols *must* be non-empty.

Reply:
I'd say let's leave this unchanged.

Reply:
Specifically, the behavior.

Reply:
sounds good to me.